### ISSUE TITLE
Cannot use after sucesfull install - cannot load extlz4.so

### ISSUE THREAD

**Issue by johnny-b2** [dearblue/ruby-extlz4#2]
Hi,
I'm trying tu use this gem for ruby-kafka client. I installed gem sucesfully.
I'm using rhscl/ruby-25-rhel7 image.
```
$ gem install extlz4

/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/HISTORY.ja.md
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/LICENSE
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/README.md
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/Rakefile
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/bin/extlz4
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/INSTALL
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/LICENSE
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/NEWS
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/README.md
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/LICENSE
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/README.md
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/liblz4.pc.in
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4frame.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4frame.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4frame_static.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4hc.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/lz4hc.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/xxhash.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/contrib/lz4/lib/xxhash.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/examples/frameapi.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/blockapi.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/depend
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/extconf.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/extlz4.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/extlz4.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/frameapi.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/hashargs.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/hashargs.h
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/lz4_amalgam.c
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/gemstub.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4/compat.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4/fix-0.1bug.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4/oldstream.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4/version.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/test/common.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/test/test_blockapi.rb
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/test/test_frameapi.rb
Building native extensions. This could take a while...
current directory: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext
/opt/rh/rh-ruby25/root/usr/bin/ruby -r ./siteconf20200603-1029-1jx5dpo.rb extconf.rb
checking for whether -I$(srcdir)/../contrib/lz4/lib is accepted as CPPFLAGS... yes
checking for whether -fvisibility=hidden is accepted as CFLAGS... yes
creating Makefile
current directory: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext
make "DESTDIR=" clean
rm -f 
rm -f extlz4.so  *.o  *.bak mkmf.log .*.time
current directory: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext
make "DESTDIR="
gcc -I. -I/opt/rh/rh-ruby25/root/usr/include -I/opt/rh/rh-ruby25/root/usr/include/ruby/backward -I/opt/rh/rh-ruby25/root/usr/include -I. '-DRBEXT_API=__attribute__ ((visibility("default")))' -DRBEXT_VISIBILITY=1    -I./../contrib/lz4/lib -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fvisibility=hidden -m64 -o blockapi.o -c blockapi.c
�[91mblockapi.c: In function 'aux_LZ4_compress_fast_continue':
blockapi.c:35:12: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
     return (int)aux_thread_call_without_gvl(
            ^
blockapi.c: In function 'aux_LZ4_compressHC_continue':
blockapi.c:57:12: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
     return (int)aux_thread_call_without_gvl(
            ^
blockapi.c: In function 'aux_LZ4_decompress_safe_continue':
blockapi.c:79:12: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
     return (int)aux_thread_call_without_gvl(
            ^
�[0mgcc -I. -I/opt/rh/rh-ruby25/root/usr/include -I/opt/rh/rh-ruby25/root/usr/include/ruby/backward -I/opt/rh/rh-ruby25/root/usr/include -I. '-DRBEXT_API=__attribute__ ((visibility("default")))' -DRBEXT_VISIBILITY=1    -I./../contrib/lz4/lib -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fvisibility=hidden -m64 -o extlz4.o -c extlz4.c
gcc -I. -I/opt/rh/rh-ruby25/root/usr/include -I/opt/rh/rh-ruby25/root/usr/include/ruby/backward -I/opt/rh/rh-ruby25/root/usr/include -I. '-DRBEXT_API=__attribute__ ((visibility("default")))' -DRBEXT_VISIBILITY=1    -I./../contrib/lz4/lib -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fvisibility=hidden -m64 -o frameapi.o -c frameapi.c
gcc -I. -I/opt/rh/rh-ruby25/root/usr/include -I/opt/rh/rh-ruby25/root/usr/include/ruby/backward -I/opt/rh/rh-ruby25/root/usr/include -I. '-DRBEXT_API=__attribute__ ((visibility("default")))' -DRBEXT_VISIBILITY=1    -I./../contrib/lz4/lib -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fvisibility=hidden -m64 -o hashargs.o -c hashargs.c
gcc -I. -I/opt/rh/rh-ruby25/root/usr/include -I/opt/rh/rh-ruby25/root/usr/include/ruby/backward -I/opt/rh/rh-ruby25/root/usr/include -I. '-DRBEXT_API=__attribute__ ((visibility("default")))' -DRBEXT_VISIBILITY=1    -I./../contrib/lz4/lib -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -fvisibility=hidden -m64 -o lz4_amalgam.o -c lz4_amalgam.c
rm -f extlz4.so
gcc -shared -o extlz4.so blockapi.o extlz4.o frameapi.o hashargs.o lz4_amalgam.o -L. -L/opt/rh/rh-ruby25/root/usr/lib64 -L. -Wl,-z,relro -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,-z,relro  -m64  -m64  -lruby  -lpthread -ldl -lcrypt -lm   -lc
current directory: /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext
make "DESTDIR=" install
/usr/bin/mkdir -p . ./.gem.20200603-1029-b6hxzf/2.5
exit > .sitearchdir.-.2.5.time
/usr/bin/install -c -m 0755 extlz4.so ./.gem.20200603-1029-b6hxzf/2.5
/opt/rh/rh-ruby25/root/usr/local/bin/extlz4
Successfully installed extlz4-0.3
```

But when I try to load extlz4 or run the lz4 binary I get an error about missing extlz4.so file:
```
sh-4.2# /opt/rh/rh-ruby25/root/usr/local/bin/extlz4
Traceback (most recent call last):
	6: from /opt/rh/rh-ruby25/root/usr/local/bin/extlz4:23:in `<main>'
	5: from /opt/rh/rh-ruby25/root/usr/local/bin/extlz4:23:in `load'
	4: from /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/bin/extlz4:7:in `<top (required)>'
	3: from /opt/rh/rh-ruby25/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
	2: from /opt/rh/rh-ruby25/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
	1: from /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4.rb:7:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4.rb:7:in `require_relative': cannot load such file -- /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/2.5/extlz4.so (LoadError)
```

Any help will be appriciated. Thanks

**Comment by johnny-b2** [comment]
I just found that extlz4.so file is in ext directory:
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/extlz4.so 
When I copy the file to 
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/2.5/extlz4.so 
Then I can run the extlz4 binary successfully.

How can I do the fix to make symlink or mv the file to the expected location or how can I change the location where to find the file? I have to done that in .gem file prior to installing gem (not after gem install command), because I'm installing the gem during the image build process.

**Comment by dearblue** [comment]
Hi,
I could not reproduce it on my FreeBSD 12.1 system (`gem25 install --user-install` and `gem25 install --no-user-install` in both cases).

If you look at the third line from the bottom of the build process log you were shown, `extlz4.so` The file seems to have been installed successfully (`./.gem.20200603-1029-b6hxzf` is a temporary directory, which is gem provides).
I think gem will move the file to the right place after this...

**Comment by johnny-b2** [comment]
Thanks for the answer. I tested it on Fedora 32 / CentOS 7.8 / RHEL 7.8.

I found that after gem installation there is `extlz4.so` on two loccations:
```
sh-4.2# find / -name extlz4.so
/opt/rh/rh-ruby25/root/usr/local/lib64/gems/ruby/extlz4-0.3/2.5/extlz4.so
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/ext/extlz4.so
```

But the `extlz4` binary is looking/expecting the library in different location:
```
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/2.5/extlz4.so
```

As show in the output:
```
sh-4.2# /opt/rh/rh-ruby25/root/usr/local/bin/extlz4
Traceback (most recent call last):
	6: from /opt/rh/rh-ruby25/root/usr/local/bin/extlz4:23:in `<main>'
	5: from /opt/rh/rh-ruby25/root/usr/local/bin/extlz4:23:in `load'
	4: from /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/bin/extlz4:7:in `<top (required)>'
	3: from /opt/rh/rh-ruby25/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
	2: from /opt/rh/rh-ruby25/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
	1: from /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4.rb:7:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/extlz4.rb:7:in `require_relative': cannot load such file -- /opt/rh/rh-ruby25/root/usr/local/share/gems/gems/extlz4-0.3/lib/2.5/extlz4.so (LoadError)
```

When I copy the file to the expected location or make a symlink, then the program works.

Is it possible to manage installation of the gem to works on both the BSD and the Linux platforms?

**Comment by dearblue** [comment]
I figured out the directory structure.

https://github.com/dearblue/ruby-extlz4/blob/3aec15ad7c01e81a8ed36d3b9dcf3b65350d205f/lib/extlz4.rb#L7

Setting `require_relative` to `require` seems to solve the problem.

> Is it possible to manage installation of the gem to works on both the BSD and the Linux platforms?

Of course it is possible.
Please wait a few days for the release.

**Comment by johnny-b2** [comment]
Thanks for quick reply.
Awsome. Looking forward for patched release.
Thanks in advance.

**Comment by dearblue** [comment]
Thanks for the very valuable report.
I have released extlz4-0.3.1, which fixes the problem.
I have also confirmed that the problem has been fixed on fedora 32.
If you are still having problems, please let me know and I would be very grateful.

**Comment by johnny-b2** [comment]
I can confirm extlz4-0.3.1 fixes the problem - tested on RHEL 7.
Thanks again for quick fix.
